### ISSUE TITLE
BSV Peers freezing and struggling to stay in sync with blockchain

### ISSUE THREAD

**Issue by jamesgithub1** [bitcoin-sv/bitcoin-sv#308]
### Describe the issue
As of roughly Wednesday last week (13th of December) my BitcoinSV peers have become unresponsive and are failing to maintain sufficient sync speed to keep up with the blockchain. I have also noticed that they are failing to consistently maintain connections to other peers with the number of connections being from 0 to 3.

These were working as expected before last week.

I have double checked and infrastructure and network appear to all be working as expected still so I can't work out the cause of the issues.

There is also a large increase in the amount of memory being used for each peer and a noticeable increase in network activity I can see.

It'd be great to know:
1. If there is any known cause for this. Are other peers having this issue
2. Did any change in the network activity happen last week which might explain this
3. Any suggestions for resolving the issue

#### If so, please list the steps to reproduce below:
1. The issue is ongoing even after restarting the Peers over a period of the past few days.

### Expected behaviour
1. Peers can sync with blockchain
4. Peers are responsive to RPC requests
5. Peers can maintain connections with other network peers.

### Actual behaviour
1. Peers are struggling to sync with blockchain. The block the peer has reached is maintaining and even falling further behind the rest of the network. 
2. Sometimes RPC requests are responsive however frequently even under low load they become completely unresponsive for periods of a few hours.
3. The peers are not reliably maintaining what i believe is an expected number of connection with network peers

### What version of bitcoin-sv are you using?
Two impacted peers
Peer 1 - 1.0.14
Peer 2 - 1.0.11

Both peers behave similarly

### Machine specs:
- OS: Centos 8
- CPU: 20CPU
- RAM: 64GB RAM with 64GB Swap
- Disk size: 32TB  
- Disk Type (HD/SDD): SSD
- Network - 1Gbit+ (up and down)

### Any extra information that might be useful in the debugging process.
This is normally the contents of a `bitcoind.log` or `config.log` file. Raw text or a link to a pastebin type site are preferred.

```
Conf
server=1
whitelist=0.0.0.0/0
txindex=1
addressindex=1
timestampindex=1
spentindex=1

zmqpubhashblock=tcp://*:28332
#zmqpubhashtx=tcp://*:28332

# RPC settings
rpcallowip=0.0.0.0/0
rpcport=8080
rpcuser=bitcoin
rpcpassword=PASSWORD
rpcworkqueue=800
rpcthreads=16
rpcclienttimeout=150
rpcservertimeout=150

uacomment=bitcoin-sv
onlynet=ipv4
dbcache=8192
listenonion=0
regtest=0
testnet=0
rest=1

maxconnections=80

# Required Consensus Rules for Genesis
excessiveblocksize=10000000000         #10GB
maxstackmemoryusageconsensus=100000000 #100MB

# Mining
minminingtxfee=0.00000500

# Allow more time for 4gb blocks - 20 mins (200% of 10 mins)
blockdownloadtimeoutbasepercent=200
```

**Comment by mickmight** [comment]
Thank you for submitting this report. My first observation is that you are not using the latest version of the BSV Node software. 1.0.16 is the latest version.

Significant memory optimizations were released in 1.0.15 and there were also thread management optimizations in 1.0.16 so it is worth considering taking advantage of these improvements. 

For the moment though, there is a diagnostic test script you can run. The diagnostic test and corresponding readme file reside [here](https://github.com/bitcoin-sv/bitcoin-sv/tree/master/contrib/diagnostic). It would help the technical team if you can please run this diagnostic script and share the results this us, either publicly on GitHub or privately via the Bitcoin BSV Association.
Thanks for your error report and I look forward to seeing your diagnostic script execution results. 
BSV Node team.

**Comment by jamesgithub1** [comment]
Hi. Thanks for the response. Below you can find the output of the script
`{
"system_info": {
"Date": "2023-12-18 12:09:27",
"Manufacturer": "Red Hat",
"Product Name": "OpenStack Compute",
"Version": "16.0.3-2.el7",
"Operating System": "  Operating System: CentOS Linux 8 (Core)",
"Kernel": "4.18.0-147.3.1.el8_1.x86_64",
"Architecture": "x86_64",
"Processor Name": "Intel Core Processor (Haswell, no TSX)",
"IP addresses": "10.161.123.91 10.165.54.120 ",
"Memory usage": "31735/64160MB (49.46%)",
"Disk usage": "6/25GB (28%)",
"Current CPU usage": "0.6%",
"Overal CPU usage": "1.56833%",
"RPC executed on": "local machine" ,
"node_processes":  [{"user": "user", "pid": "1636", "cpu": "26.9", "mem": "48.5", "command": "XXXXXX/bin/bitcoind"}]
},
"RPC" :{
"getbestblockhash": "ERROR",
"getblockchaininfo":"ERROR",
"getdifficulty": "ERROR",
"getmempoolinfo":"ERROR",
"getrawnonfinalmempool": [
],
"getmemoryinfo": {
  "locked": {
    "used": 64032,
    "free": 198112,
    "total": 262144,
    "locked": 262144,
    "chunks_used": 2001,
    "chunks_free": 775
  },
  "preloading": {
    "chainStateCached": 12.30146891110795
  }
},
"uptime": 75912,
"getmininginfo": {
  "blocks": 823026,
  "currentblocksize": 0,
  "currentblocktx": 0,
  "difficulty": 72332442795.56525,
  "errors": "Warning: The network does not appear to agree with the local blockchain! Still waiting for block data for more details. We are in the startup process (e.g. Initial block download or reindex), this might be reason for disagreement.",
  "networkhashps": 5.664622380717866e+17,
  "pooledtx": 0,
  "chain": "main"
},
"getconnectioncount" :3,
"getexcessiveblock": {
  "excessiveBlockSize": 10000000000
},
"getnettotals": {
  "totalbytesrecv": 219605020,
  "totalbytessent": 190066,
  "timemillis": 1702923677104,
  "uploadtarget": {
    "timeframe": 86400,
    "target": 0,
    "target_reached": false,
    "serve_historical_blocks": true,
    "bytes_left_in_cycle": 0,
    "time_left_in_cycle": 0
  }
},
"getnetworkinfo": {
  "version": 101001400,
  "subversion": "/Bitcoin SV:1.0.14(bitcoin-sv)/",
  "protocolversion": 70016,
  "localservices": "0000000000000021",
  "localrelay": true,
  "timeoffset": -8,
  "txnpropagationfreq": 250,
  "txnpropagationqlen": 0,
  "networkactive": true,
  "connections": 3,
  "addresscount": 41935,
  "streampolicies": "BlockPriority,Default",
  "networks": [
    {
      "name": "ipv4",
      "limited": false,
      "reachable": true,
      "proxy": "",
      "proxy_randomize_credentials": false
    },
    {
      "name": "ipv6",
      "limited": true,
      "reachable": false,
      "proxy": "",
      "proxy_randomize_credentials": false
    }
  ],
  "relayfee": 0.00000000,
  "minconsolidationfactor": 20,
  "maxconsolidationinputscriptsize": 150,
  "minconfconsolidationinput": 6,
  "minconsolidationinputmaturity": 6,
  "acceptnonstdconsolidationinput": false,
  "localaddresses": [
  ],
  "warnings": "Warning: The network does not appear to agree with the local blockchain! Still waiting for block data for more details. We are in the startup process (e.g. Initial block download or reindex), this might be reason for disagreement."
},
"getpeerinfo": [
  {
    "id": 1058,
    "addr": "65.108.143.94:8333",
    "addrlocal": "204.50.64.2:23643",
    "services": "0000000000000021",
    "relaytxes": true,
    "lastsend": 1702923677,
    "lastrecv": 1702923677,
    "sendsize": 0,
    "recvsize": 0,
    "sendmemory": 0,
    "pausesend": false,
    "unpausesend": false,
    "bytessent": 564,
    "bytesrecv": 111739,
    "avgrecvbw": 398,
    "associd": "241c1d4e-5f52-47e4-9d8d-5a5572f43fc1",
    "streampolicy": "BlockPriority",
    "streams": [
      {
        "streamtype": "GENERAL",
        "lastsend": 1702923378,
        "lastrecv": 1702923677,
        "bytessent": 443,
        "bytesrecv": 111603,
        "sendsize": 0,
        "recvsize": 0,
        "sendmemory": 0,
        "spotrecvbw": 1482,
        "minuterecvbw": 796,
        "pauserecv": false
      },
      {
        "streamtype": "DATA1",
        "lastsend": 1702923677,
        "lastrecv": 1702923498,
        "bytessent": 121,
        "bytesrecv": 136,
        "sendsize": 0,
        "recvsize": 0,
        "sendmemory": 0,
        "spotrecvbw": 0,
        "minuterecvbw": 0,
        "pauserecv": false
      }
    ],
    "authconn": false,
    "conntime": 1702923321,
    "timeoffset": -57,
    "pingtime": 0.44695,
    "minping": 0.44695,
    "pingwait": 0.038024,
    "version": 70016,
    "subver": "/Bitcoin SV:1.0.14/",
    "inbound": false,
    "addnode": false,
    "startingheight": 823180,
    "txninvsize": 0,
    "banscore": 0,
    "synced_headers": -1,
    "synced_blocks": -1,
    "inflight": [
    ],
    "whitelisted": false,
    "bytessent_per_msg": {
      "authch": 64,
      "createstrm": 57,
      "getaddr": 24,
      "ping": 64,
      "pong": 64,
      "protoconf": 51,
      "sendcmpct": 33,
      "sendheaders": 24,
      "verack": 24,
      "version": 159
    },
    "bytesrecv_per_msg": {
      "addr": 30082,
      "authch": 64,
      "inv": 81175,
      "ping": 64,
      "pong": 32,
      "protoconf": 51,
      "sendcmpct": 33,
      "sendheaders": 24,
      "streamack": 43,
      "verack": 24,
      "version": 147
    }
  },
  {
    "id": 1063,
    "addr": "54.249.171.1:8333",
    "addrlocal": "204.50.64.2:41906",
    "services": "0000000000000021",
    "relaytxes": true,
    "lastsend": 1702923677,
    "lastrecv": 1702923620,
    "sendsize": 0,
    "recvsize": 0,
    "sendmemory": 0,
    "pausesend": false,
    "unpausesend": false,
    "bytessent": 411,
    "bytesrecv": 222,
    "avgrecvbw": 8,
    "associd": "9a8ac31c-c8f4-4e4b-beef-7e9cc63cc246",
    "streampolicy": "BlockPriority",
    "streams": [
      {
        "streamtype": "GENERAL",
        "lastsend": 1702923677,
        "lastrecv": 1702923620,
        "bytessent": 411,
        "bytesrecv": 222,
        "sendsize": 0,
        "recvsize": 0,
        "sendmemory": 0,
        "spotrecvbw": 0,
        "minuterecvbw": 8,
        "pauserecv": false
      }
    ],
    "authconn": false,
    "conntime": 1702923619,
    "timeoffset": -502,
    "pingwait": 0.0367,
    "version": 70016,
    "subver": "/Bitcoin SV:1.0.16/",
    "inbound": false,
    "addnode": false,
    "startingheight": 823182,
    "txninvsize": 0,
    "banscore": 0,
    "synced_headers": -1,
    "synced_blocks": -1,
    "inflight": [
    ],
    "whitelisted": false,
    "bytessent_per_msg": {
      "authch": 64,
      "getaddr": 24,
      "ping": 32,
      "protoconf": 51,
      "sendcmpct": 33,
      "sendheaders": 24,
      "verack": 24,
      "version": 159
    },
    "bytesrecv_per_msg": {
      "protoconf": 51,
      "verack": 24,
      "version": 147
    }
  },
  {
    "id": 1064,
    "addr": "78.110.160.26:8333",
    "services": "0000000000000000",
    "relaytxes": false,
    "lastsend": 1702923675,
    "lastrecv": 0,
    "sendsize": 0,
    "recvsize": 0,
    "sendmemory": 0,
    "pausesend": false,
    "unpausesend": false,
    "bytessent": 159,
    "bytesrecv": 0,
    "avgrecvbw": 0,
    "associd": "9cb1fa12-d492-4423-be6b-3d8b5082a730",
    "streampolicy": "Default",
    "streams": [
      {
        "streamtype": "GENERAL",
        "lastsend": 1702923675,
        "lastrecv": 0,
        "bytessent": 159,
        "bytesrecv": 0,
        "sendsize": 0,
        "recvsize": 0,
        "sendmemory": 0,
        "spotrecvbw": 0,
        "minuterecvbw": 0,
        "pauserecv": false
      }
    ],
    "authconn": false,
    "conntime": 1702923675,
    "timeoffset": 0,
    "version": 0,
    "subver": "",
    "inbound": false,
    "addnode": false,
    "startingheight": -1,
    "txninvsize": 0,
    "banscore": 0,
    "synced_headers": -1,
    "synced_blocks": -1,
    "inflight": [
    ],
    "whitelisted": false,
    "bytessent_per_msg": {
      "version": 159
    },
    "bytesrecv_per_msg": {
    }
  }
],
"listbanned": [
  {
    "address": "3.11.11.106/32",
    "banned_until": 1703002439,
    "ban_created": 1702916039,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "3.21.162.186/32",
    "banned_until": 1702958551,
    "ban_created": 1702872151,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "3.65.143.234/32",
    "banned_until": 1702958551,
    "ban_created": 1702872151,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "3.70.182.112/32",
    "banned_until": 1702949552,
    "ban_created": 1702863152,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "3.97.225.48/32",
    "banned_until": 1702959138,
    "ban_created": 1702872738,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "3.126.128.168/32",
    "banned_until": 1703001590,
    "ban_created": 1702915190,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "3.223.106.89/32",
    "banned_until": 1703000171,
    "ban_created": 1702913771,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "3.228.193.128/32",
    "banned_until": 1702959459,
    "ban_created": 1702873059,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "13.213.28.214/32",
    "banned_until": 1702950005,
    "ban_created": 1702863605,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "18.139.1.192/32",
    "banned_until": 1702980108,
    "ban_created": 1702893708,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "35.77.137.101/32",
    "banned_until": 1702980814,
    "ban_created": 1702894414,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "35.209.186.110/32",
    "banned_until": 1702925298,
    "ban_created": 1702838898,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "44.225.254.244/32",
    "banned_until": 1702949887,
    "ban_created": 1702863487,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "45.8.132.10/32",
    "banned_until": 1702950017,
    "ban_created": 1702863617,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "46.166.142.52/32",
    "banned_until": 1702982153,
    "ban_created": 1702895753,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "46.250.249.70/32",
    "banned_until": 1702949407,
    "ban_created": 1702863007,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "52.25.255.219/32",
    "banned_until": 1702990121,
    "ban_created": 1702903721,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "52.78.247.110/32",
    "banned_until": 1702970265,
    "ban_created": 1702883865,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "52.197.6.170/32",
    "banned_until": 1702970251,
    "ban_created": 1702883851,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "54.244.67.182/32",
    "banned_until": 1702968396,
    "ban_created": 1702881996,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "59.167.208.48/32",
    "banned_until": 1702958587,
    "ban_created": 1702872187,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "62.171.183.65/32",
    "banned_until": 1702989890,
    "ban_created": 1702903490,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "75.98.207.196/32",
    "banned_until": 1703007894,
    "ban_created": 1702921494,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "80.241.211.40/32",
    "banned_until": 1702957973,
    "ban_created": 1702871573,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "88.99.103.51/32",
    "banned_until": 1702990960,
    "ban_created": 1702904560,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "89.179.247.236/32",
    "banned_until": 1702924370,
    "ban_created": 1702837970,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "91.237.249.24/32",
    "banned_until": 1703001342,
    "ban_created": 1702914942,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "93.95.228.200/32",
    "banned_until": 1702926835,
    "ban_created": 1702840435,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "116.203.134.8/32",
    "banned_until": 1703000742,
    "ban_created": 1702914342,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "141.98.219.144/32",
    "banned_until": 1702989890,
    "ban_created": 1702903490,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "144.91.65.82/32",
    "banned_until": 1703001227,
    "ban_created": 1702914827,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "144.91.71.139/32",
    "banned_until": 1702988964,
    "ban_created": 1702902564,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "144.202.91.233/32",
    "banned_until": 1702925296,
    "ban_created": 1702838896,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "161.97.72.190/32",
    "banned_until": 1702999448,
    "ban_created": 1702913048,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "162.242.168.55/32",
    "banned_until": 1703001585,
    "ban_created": 1702915185,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "162.242.168.57/32",
    "banned_until": 1703000921,
    "ban_created": 1702914521,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "163.172.50.16/32",
    "banned_until": 1702980165,
    "ban_created": 1702893765,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "163.172.86.28/32",
    "banned_until": 1702981218,
    "ban_created": 1702894818,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "167.86.89.162/32",
    "banned_until": 1703000337,
    "ban_created": 1702913937,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "168.119.86.3/32",
    "banned_until": 1702968640,
    "ban_created": 1702882240,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "168.119.86.11/32",
    "banned_until": 1703009778,
    "ban_created": 1702923378,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "168.119.162.233/32",
    "banned_until": 1702999401,
    "ban_created": 1702913001,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "173.249.58.36/32",
    "banned_until": 1702970777,
    "ban_created": 1702884377,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "178.18.250.227/32",
    "banned_until": 1702999441,
    "ban_created": 1702913041,
    "ban_reason": "node misbehaving"
  },
  {
    "address": "194.14.247.123/32",
    "banned_until": 1702949588,
    "ban_created": 1702863188,
    "ban_reason": "node misbehaving"
  }
],
"dumpparameters": [
  "addressindex=1",
  "blockdownloadtimeoutbasepercent=200",
  "conf=/sites/bitu-bitcoinsv.internal/config/bitcoin.conf",
  "daemon",
  "datadir=/bitcoin",
  "dbcache=8192",
  "excessiveblocksize=10000000000",
  "listenonion=0",
  "maxconnections=80",
  "maxstackmemoryusageconsensus=100000000",
  "minminingtxfee=0.00000500",
  "onlynet=ipv4",
  "pid=/run/bitcoind/bitcoind.pid",
  "regtest=0",
  "rest=1",
  "rpcallowip=0.0.0.0/0",
  "rpcclienttimeout=150",
  "rpcport=8080",
  "rpcservertimeout=150",
  "rpcthreads=16",
  "rpcworkqueue=800",
  "server=1",
  "spentindex=1",
  "testnet=0",
  "timestampindex=1",
  "txindex=1",
  "uacomment=bitcoin-sv",
  "wallet=wallet.dat",
  "whitelist=0.0.0.0/0",
  "whitelistrelay=1",
  "zmqpubhashblock=tcp://*:28332"
],
"getsettings": {
  "excessiveblocksize": 10000000000,
  "blockmaxsize": 128000000,
  "maxtxsizepolicy": 10000000,
  "maxorphantxsize": 1000000000,
  "datacarriersize": 4294967295,
  "maxscriptsizepolicy": 500000,
  "maxopsperscriptpolicy": 4294967295,
  "maxscriptnumlengthpolicy": 10000,
  "maxpubkeyspermultisigpolicy": 4294967295,
  "maxtxsigopscountspolicy": 4294967295,
  "maxstackmemoryusagepolicy": 100000000,
  "maxstackmemoryusageconsensus": 100000000,
  "limitancestorcount": 10000,
  "limitcpfpgroupmemberscount": 25,
  "maxmempool": 1000000000,
  "maxmempoolsizedisk": 0,
  "mempoolmaxpercentcpfp": 10,
  "acceptnonstdoutputs": true,
  "datacarrier": true,
  "minminingtxfee": 0.00000500,
  "maxstdtxvalidationduration": 3,
  "maxnonstdtxvalidationduration": 1000,
  "maxtxchainvalidationbudget": 50,
  "validationclockcpu": true,
  "minconsolidationfactor": 20,
  "maxconsolidationinputscriptsize": 150,
  "minconfconsolidationinput": 6,
  "minconsolidationinputmaturity": 6,
  "acceptnonstdconsolidationinput": false
}
}
}`

**Comment by mickmight** [comment]
Thanks, the team will review this and provide feedback.

**Comment by mickmight** [comment]
The team have had a look at your diagnostics report, James. This was insightful, and thanks for sharing it. 

There are two areas of investigation. One area investigates the responsiveness issue your node encountered. The second area is looking at a small number of peer connections your node has. 

To progress the first area, the team would like to know your typical usage patterns for the RPC calls you make on your nodes. What do you call the most and how frequently, what typical response times do you see? 

Ideally the team could see your bitcoind.log file (from 13th and 18th Dec) and would welcome these to be shared privately with the team. Also the stderr output from the failed RPC calls your node encountered would be useful to see if you have them. Can you also confirm please, that your bitcoin.conf settings were the same on the 13th and 18th of December?

Regarding the second area of investigation, the team are performing some tests with a similar network topology of between 0 and 3 connect peers.

**Comment by ScottMcDermid** [comment]
We have been experiencing a similar behavior. Our node is struggling to keep up with the latest block since around last week.

### What version of bitcoin-sv are you using?

Node Version: 1.0.14

### Machine specs:

    OS: Almalinux 8.9
    CPU: AMD Ryzen 7 5800X 8-Core Processor
    RAM: 96GiB DDR4 2666MHz
    Disk size: 20TB
    Disk Type (HD/SDD): SSD

### Any extra information that might be useful in the debugging process.

```
##
## bitcoin.conf configuration file. Lines beginning with # are comments.
##
 
# Network-related settings:

datadir=/media/bitcoin-sv/data

maxmempool=25000
dbcache=8192

excessiveblocksize=10000000000
maxstackmemoryusageconsensus=8000000000


txindex=1

# Maximum number of inbound+outbound connections.
#maxconnections=24

maxlimitertxfee=0

#preload mempool
preload=1

#
# JSON-RPC options (for controlling a running Bitcoin/bitcoind process)
#

server=1

rest=1

rpcuser=USER
rpcpassword=PASSWORD

rpcallowip=0.0.0.0/0

port=9094
rpcport=10006

rpcworkqueue=800
rpcthreads=16

# How many seconds bitcoin will wait for a complete RPC HTTP request.
# after the HTTP connection is established. 
# DOESN'T WORK IN BitcoinUnlimited as of 1.5.0.0
rpcclienttimeout=120

rpcservertimeout=180

# Create transactions that have enough fees (or priority) so they are likely to begin confirmation within n blocks (default: 1).
# This setting is over-ridden by the -paytxfee option.
txconfirmtarget=6

# minimum mining fee (required for 1.0.11)
minminingtxfee=0.000005

# Miscellaneous options

# Pre-generate this many public/private key pairs, so wallet backups will be valid for
# both prior transactions and several dozen future transactions.
keypool=0

# Pay an optional transaction fee every time you send bitcoins.  Transactions with fees
# are more likely than free transactions to be included in generated blocks, so may
# be validated sooner.
#paytxfee=0.00

addnode=seed.bitcoinsv.io
addnode=seed.cascharia.com
addnode=seed.satoshisvision.network
```

**Comment by mickmight** [comment]
@ScottMcDermid thanks for reporting this. I shall ask the development team to analyse this. I did notice that your spec is below the recommended minimum. https://node.bitcoinsv.io/sv-node/system-requirements was fairly recently updated to show >= 10 cores and 128GB of memory as minimum recommended prod spec. While this may not be the root cause, I wanted to highlight my initial observation. 

Will come back to you soon when I have more. 

Regards,
Mick

**Comment by mickmight** [comment]
@ScottMcDermid one of the engineers has read through your report. It sounds similar to what the original post described. To enable to Node team to look more into this behaviour, it would be enormously helpful if you could provide:

1. The results of running the diagnostic script located [here](https://github.com/bitcoin-sv/bitcoin-sv/tree/master/contrib/diagnostic).
2. Typical usage patterns for the RPC calls you make on your node(s). What do you call the most and how frequently, what typical response times do you see?
3. Bitcoind.log for a period where your node was unresponsive.
4. stderr output showing any failed RPC calls your node encountered.

Thanks,
Mick

**Comment by ScottMcDermid** [comment]
Thanks for your help!

> https://node.bitcoinsv.io/sv-node/system-requirements was fairly recently updated to show >= 10 cores and 128GB of memory as minimum recommended prod spec. While this may not be the root cause, I wanted to highlight my initial observation.

I introduced another 160gb of swap. We effectively have 256gb now.

> 1. The results of running the diagnostic script located [here](https://github.com/bitcoin-sv/bitcoin-sv/tree/master/contrib/diagnostic).

As of last night, we're struggling to even start the node. If we succeed in starting it, I will follow up with the diagnostic output.

> 2. Typical usage patterns for the RPC calls you make on your node(s). What do you call the most and how frequently, what typical response times do you see?

Typically, we run a service that calls `getblockhash` and `getblock` quite heavily. This service takes between 1-5 minutes to complete.

> 3. Bitcoind.log for a period where your node was unresponsive.

I included 2 startup attempts. The first one succeeded, and runs for about 15 mins, and the second fails to start at all, which seems to be the more common outcome.

[bitcoind.log](https://github.com/bitcoin-sv/bitcoin-sv/files/13746942/bitcoind.log)

> 4. stderr output showing any failed RPC calls your node encountered.

We can't recall what the RPC error message was while the node was running. At this point, we're only seeing the usual startup errors, as it rarely seems to start.
```
error code: -28
error message:
Verifying blocks...
```

**Comment by ScottMcDermid** [comment]
I was able to gather some data from the diagnostic script before the node crashed:

[report.json](https://github.com/bitcoin-sv/bitcoin-sv/files/13747327/report.json)

**Comment by mickmight** [comment]
Hi Scott, thank you for sharing this additional information. The service you are running (that makes calls to the `getBlockHash `and `getBlock` RPC functions "quite heavily" could explain why your node is so busy.

If the blocks are large (such as 4GB blocks, which have been seen recently on the network), the `getBlock` RPC will take longer to execute. During this time, it will consume one CPU core and may also use more disk I/O. These consumed resources are then unavailable for normal node operations (e.g. tx/ block processing) and as a result, your node will take longer to keep pace with the blockchain. If your service makes numerous of these RPC calls in parallel, the impact is proportionally larger.

We are aware that high demands on the RPC interface poses a challenge and are considering ways to alleviate this in future releases of the node software.

**Additional technical analysis:**

- Binding RPC on address 0.0.0.0 port 10008 failed. These could indicate that another instance of bitcoind (or another process using the same port) is already running.
- The node also shuts down during start-up. The log file has ~90 such unsuccessful start attempts, all on 2023-12-21.
- Log file bitcoind.log contains a few instances of "[worker1-N-HTTPServer] Received RPC call stop()". These should have been preceded by "[worker1-N-HTTPServer] ThreadRPCServer method=stop", but they are not. 
    - This suggests that RPC logging is disabled in configuration. 
    - Consequently it is not possible to see what RPC functions were being called.
- Log file bitcoind.log shows many (80+) attempts to start bitcoind, which then stops after a few seconds. 
    - A few of those stops have been requested via RPC (see previous comment), a few at the start of the log have failed because no RPC port could be bound, and a majority of these attempts contain no reason for stopping.
    - I find the following two ways bitcoind process could be stopped regularly without a reason being seen in a log file.
        - By using SIGTERM signal.
        - By using` -stopatheight` configuration option.
    - Both suggest that stopping has external origin.
    - It is not possible to conclude anything with a reasonable certainty. 
    - A malfunctioning script/process/service used to start/stop bitcoind could cause observed behaviour.

**Other considerations:**
- To get a fuller picture of RPC usage, you could enable debug logging such:
```
debug=rpc
debug=http
```
- Mempool size could be reduced to 16GB for instance:
`maxmempool=16GB`
- Upgrading to 1.0.16 would also take advance of some of the recent optimisations that have been added to the node software.

**Comment by ScottMcDermid** [comment]
Thanks for the troubleshooting information. We discovered a service interfering with our node, and disabled it.

Our node is no longer shutting down instantly, and only experiences the original issue of struggling to sync.

We will update our settings to enable debug logs, and get back to you about our RPC usage.

> Upgrading to 1.0.16 would also take advance of some of the recent optimisations that have been added to the node software.

We're in the process of installing form source, as the existing binaries don't seem to run out-of-the-box for Almalinux 8.9. (Since version 1.0.15).

**Other observations**

- Farthest we've been able to sync is block `823492`
- Calls to `getblockhash` and `getblock` for older blocks do not respond (tested on block `823201`)

**Comment by relayking** [comment]
Issue with stability/not keeping up seems to be related to chainstate compaction (?) taking absurd amount of time. Would be nice to have some knobs around how/when it happens.

**Comment by lifengmds** [comment]
Hard to keep synchronized recently, why the size of chainstate folder has gone up to around 300G ?

**Comment by bvbfan** [comment]
> Hard to keep synchronized recently, why the size of chainstate folder has gone up to around 300G ?

It's coins db, utxo will continue to grow because of ordinals.

https://whatsonchain.com/block-stat/total_utxo_increase

**Comment by relayking** [comment]
what about leveldb compaction freezing the node tho, dont think its supposed to work that way

**Comment by amritabithi** [comment]
I recommend changing the port that your client is listening for incoming connections on, and changing your RPC port as well if you are using RPC commands.  You can change these to any port and still be able to establish connections to the network normally.  I would also recommend adding one or two trusted peers to your peers connect list to make sure that you are getting a clean entry into the network.

I first had to do this out of necessity when I had a mining pool because at the time I had BTC and BCH both running on the same server, and they would corrupt eachother's memory and eventually cause an out of memory exception from a memory leak caused by a packet failing to process correctly, which will keep building up until it crashes the client.  If you are running your node on the default port 8333 you are sure to eventually have an errant node from another blockchain network connect to you and send you invalid packets, and so the solution is to listen on a non-default port.  This will not break anything, and peering will work as normal.

I have been forced to do this on a node I ran at home also even without other nodes running alongside it, because sometimes you will get a persistent node trying to establish peer connections to you or relay packets to you incorrectly and whatever isn't handled in that case by the software eventually exhausts the RAM memory and you will see the OOM / Out Of Memory error typically when the crash happens, but it may just exit with no logging.

I have put some sample bitcoin.conf file lines below which show the directives to place in your bitcoin.conf file to change your listening port, and a few other settings that I would adjust but are not necessary.

You can find other healthy BSV nodes by browsing the peers shown at WhatsOnChain here:
[https://whatsonchain.com/peers](https://whatsonchain.com/peers)

And this page can be helpful for editing the bitcoin.conf file and exploring other settings to change, it is designed for core but can be used just as well:
[https://jlopp.github.io/bitcoin-core-config-generator/](https://jlopp.github.io/bitcoin-core-config-generator/)

-------------------------------------------------------------

**The important line is:
\# Listen for incoming connections on non-default port ( I chose 8999 as an example, can be anything you wish).
port=8999**

-------------------------------------------------------------

```
# This config should be placed in following path:
# ~/.bitcoin/bitcoin.conf

# [network]
# Listen for incoming connections on non-default port.
port=8999

# Add a node IP address to connect to and attempt to keep the connection open. This option can be set multiple times.
# I put a random IP here, you will want to get a real node's IP which you can find here - https://whatsonchain.com/peers
addnode=11.22.33.44:8333

# Connect to a node (IP address) to retrieve peer addresses, then disconnect.
# This is a fake IP address as an example, you will want to choose a real one from here - https://whatsonchain.com/peers
seednode=222.333.444.555

# [rpc]
# Listen for JSON-RPC connections on this port ( anything but 8332 / 8333 )
# You can use whatever port you like, here I changed it from '8332' to '8992' but can be any port you wish.
rpcport=8992

# Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24). This option can be specified multiple times.
rpcallowip=127.0.0.1
```

**Comment by richmills3** [comment]
It seems this conversation has petered out. If there are still open questions feel free to open a new issue.
